### DISCUSSION TITLE
PSS code issue

### DISCUSSION THREAD

**Discussion by jinningwang** [CURENT/andes#410] [Internal Design] [answered]
Hello Hantao,

Dr. Xin Fang and I found that the PSS implementation right now might have an issue with the connection status ``u``.
``u`` can shut down the output but will not impact the input. It means a PSS is still calculated with an input signal even turned off. This might result in an unexpected TDS crash.
Do we need to take care of this, or might we need a more systematic check for connection status?

The code from version ``1.8.5.post19+g549ba98a`` is attached below for quick reference:
PSSBase:
```
        ... ...
        # from Exciter
        self.vi = ExtAlgeb(model='Exciter', src='vi', indexer=self.avr, tex_name='v_i',
                           info='Exciter input voltage',
                           e_str='u * vsout',
                           ename='Vi',
                           tex_ename='V_i',
                           is_input=True,
                           )

        self.vsout = Algeb(info='PSS output voltage to exciter',
                           tex_name='v_{sout}',
                           is_output=True,
                           )  # `self.vsout.e_str` to be provided by specific models
```

IEEEST:
```
        ... ...
        self.Vss = Algeb(tex_name='V_{ss}', info='Voltage output before output limiter',
                         e_str='VLIM_zi * WO_y + VLIM_zu * LSMAX + VLIM_zl * LSMIN - Vss')

        self.OLIM = Limiter(u=self.v, lower=self.VCLr, upper=self.VCUr, info='output limiter')

        self.vsout.e_str = 'OLIM_zi * Vss - vsout'
```

**Comment by cuihantao** [comment, marked as answer]
The best way to solve it is to provide one interface for turning off a generator. The implementation will then check all connected devices and turn that off. This might need a refactor of the current code. 

Did you observe a case where this leads to a crash? I know the value keeps getting computed. Even if the exciter signal gets changing, the generator output will stay at zero.

**Reply by XinFang88** [reply]
I have a case running into this issue. I used the IEEE 39 Bus system for the ED+TDS cosimulation. When the simulation ran second by second, the simulation crashed after several seconds. But ANDES can simulate a more extended period standalone.

**Reply by cuihantao** [reply]
Check if the latest version is being used? I remember fixing a similar issue a few months back.

**Reply by XinFang88** [reply]
My ANDES version is 1.8.4. Is this the latest version?

**Comment by cuihantao** [comment]
Yeah, this contains the fix I mentioned. Maybe we need to debug the case offline.

**Reply by XinFang88** [reply]
After removing all PSS, the simulation can run okay.

**Reply by jinningwang** [reply]
We have successfully debugged the case.
As for the code, I'll do a quick fix for the PSS connection status like what we have done for TurbinGov. Expected to finish this in one or two days.

**Reply by cuihantao** [reply]
Great. Thanks, Jinning!

**Comment by jinningwang** [comment]
Hi there, @XinFang88 @cuihantao 

A quick fix is done at PR [Fix PSS #411](https://github.com/cuihantao/andes/pull/411). Detailed test results are also included in it.
Please let me know if there is any more concern.

Regards,
Jinning
